MeesGame/Gameplay/Generator/MazeGeneratorMain.cs

```diff
@@ -0,0 +1,148 @@
+
+        private void MazeGenMain()
+        {
+            bool placeDoors = false; // Only place doors after a key has been placed.
+            bool randomNext = false; // Is set to true if the current path can't be extended
+            Tile tileToAdd;
+            int nodesDone = -1; // Makes it 0 in the first iteration.
+
+            // Add hallways until all tiles are filled (main algorithm).
+            while (nodesToDo.Count != 0)
+            {
+                nodesDone += 1;
+
+                // Continue from the last node, or take a random one
+                int position = nodesToDo.Count - 1;
+                if (randomNext || random.Next(BIGINT) < splitChance * BIGINT)
+                {
+                    // Swap the last node with another node
+                    int nodeToDoNext = random.Next(nodesToDo.Count);
+                    swap(nodesToDo, nodeToDoNext, position);
+                }
+
+                Point current = nodesToDo[position];
+                Point next;
+
+                // Make an arrray and fill it with 1-4 in a random order.
+                int[] possible = getZeroToThreeInRandomOrder();
+
+                // Set randomNext to true, it gets set to false if we can extend the hallway
+                randomNext = true;
+
+                for (int i = 0; i < 4; i++)
+                {
+                    // Decide the next point.
+                    next = decideNext(possible[i], current);
+
+                    // Check if it's in the tilefield.
+                    if (next.X < 0 || next.Y < 0 || next.X >= tiles.Columns || next.Y >= tiles.Rows) continue;
+
+                    // If it's not a walltile.
+                    if (!(tiles.GetTile(next) is WallTile))
+                    {
+                        // Making a path there would create a loop.
+                        if (random.Next(BIGINT) < loopChance * BIGINT)
+                        {
+                            // If the wall inbetween is a walltile.
+                            if (tiles.GetTile((next.X + current.X) / 2, (next.Y + current.Y) / 2) is WallTile)
+                            {
+                                // Make the loop.
+                                tileToAdd = new FloorTile();
+
+                                // Or add a door instead of making it a loop, you don't need to check if a key has been placed for this door because it's placed in a loop.
+                                if (random.Next(BIGINT) < loopDoorChance * BIGINT)
+                                {
+                                    tileToAdd = new DoorTile();
+                                }
+
+                                // Add the tile.
+                                int randNumb = random.Next(BIGINT);
+                                if (randNumb < (loopDoorChance + loopHoleChance) * BIGINT) tileToAdd = new DoorTile();
+                                if (randNumb < loopHoleChance * BIGINT) tileToAdd = new HoleTile();
```

Comment:
I've done something similar in a helper function, and fixed a calculation bug with it.